Validation/SiOuterTrackerV/plugins/OuterTrackerMonitorTrackingParticles.cc

```diff
@@ -0,0 +1,1180 @@
+    tp_d0->Fill(tmp_tp_d0);
+    tp_VtxZ->Fill(tmp_tp_VtxZ);
+
+    if (nStubTP < TP_minNStub) continue;
+    if ( !MCTruthTTTrackHandle.isValid() ) continue;
+    std::vector< edm::Ptr< TTTrack< Ref_Phase2TrackerDigi_ > > > matchedTracks = MCTruthTTTrackHandle->findTTTrackPtrs(tp_ptr);
+
+    int nMatch = 0;
+    int i_track = -1;
+    float i_chi2dof = 99999;
+
+    if (!matchedTracks.empty()) {
+      // ----------------------------------------------------------------------------------------------
+      // loop over matched L1 tracks
+      // here, "match" means tracks that can be associated to a TrackingParticle with at least one hit of at least one of its clusters
+      // https://twiki.cern.ch/twiki/bin/viewauth/CMS/SLHCTrackerTriggerSWTools#MC_truth_for_TTTrack
+      for (int it=0; it<(int)matchedTracks.size(); it++) {
+        bool tmp_trk_genuine = false;
+        if (MCTruthTTTrackHandle->isGenuine(matchedTracks.at(it))) tmp_trk_genuine = true;
+        if (!tmp_trk_genuine) continue;
+        // ----------------------------------------------------------------------------------------------
+        // further require L1 track to be (loosely) genuine, that there is only one TP matched to the track
+        // + have >= L1Tk_minNStub stubs for it to be a valid match
+        int tmp_trk_nstub = matchedTracks.at(it)->getStubRefs().size();
+        if (tmp_trk_nstub < L1Tk_minNStub) continue;
+        float dmatch_pt  = 999;
+        float dmatch_eta = 999;
+        float dmatch_phi = 999;
+        int match_id = 999;
+
+        edm::Ptr< TrackingParticle > my_tp = MCTruthTTTrackHandle->findTrackingParticlePtr(matchedTracks.at(it));
+        dmatch_pt  = fabs(my_tp->p4().pt() - tmp_tp_pt);
+        dmatch_eta = fabs(my_tp->p4().eta() - tmp_tp_eta);
+        dmatch_phi = fabs(my_tp->p4().phi() - tmp_tp_phi);
+        match_id = my_tp->pdgId();
+        float tmp_trk_chi2dof = (matchedTracks.at(it)->getChi2(L1Tk_nPar)) / (2*tmp_trk_nstub - L1Tk_nPar);
+
+        // ensure that track is uniquely matched to the TP we are looking at!
+        if (dmatch_pt<0.1 && dmatch_eta<0.1 && dmatch_phi<0.1 && tmp_tp_pdgid==match_id) {
+          nMatch++;
+          if (i_track < 0 || tmp_trk_chi2dof < i_chi2dof) {
+            i_track = it;
+            i_chi2dof = tmp_trk_chi2dof;
+          }
+        }
+
+      } // end loop over matched L1 tracks
+    } // end has at least 1 matched L1 track
+    // ----------------------------------------------------------------------------------------------
+
+    // Get information on the matched tracks
+    float tmp_matchtrk_pt   = -999;
+    float tmp_matchtrk_eta  = -999;
+    float tmp_matchtrk_phi  = -999;
+    float tmp_matchtrk_VtxZ  = -999;
+    float tmp_matchtrk_chi2 = -999;
+    float tmp_matchtrk_chi2R = -999;
+    int tmp_matchTrk_nStub  = -999;
+    float tmp_matchtrk_d0  = -999;
+
+    if (nMatch > 0) {
+      tmp_matchtrk_pt    = matchedTracks.at(i_track)->getMomentum(L1Tk_nPar).perp();
```

Comment:
use `[i_track]` (can do a bounds check separately once if needed, instead of ~10 times)